comp/core/workloadmeta/filter.go

```diff
@@ -36,38 +47,65 @@ type FilterParams struct {
 //
 // Only events of the given type will be delivered. Use EventTypeAll to collect
 // data from all the event types. EventTypeAll is the default.
-func NewFilter(filterParams *FilterParams) *Filter {
-	var kindSet map[Kind]struct{}
-	kinds := filterParams.Kinds
-	if len(kinds) > 0 {
-		kindSet = make(map[Kind]struct{})
-		for _, k := range kinds {
-			kindSet[k] = struct{}{}
-		}
+func NewFilterBuilder() *FilterBuilder {
+	return &FilterBuilder{
+		filter: &Filter{
+			source:    SourceAll,
+			eventType: EventTypeAll,
+			kinds:     map[Kind]EntityFilterFunc{},
+		},
 	}
+}
 
-	// This is enforced in the matching functions, but putting here for clarity
-	if filterParams.Source == "" {
-		filterParams.Source = SourceAll
-	}
+// Build builds the filter and returns it.
+func (fb *FilterBuilder) Build() *Filter {
+	return fb.filter
+}
 
-	return &Filter{
-		kinds:     kindSet,
-		source:    filterParams.Source,
-		eventType: filterParams.EventType,
-	}
+// AddKind adds a specific kind to the built filter.
+// The built filter will match any entity of the added kind.
+func (fb *FilterBuilder) AddKind(kind Kind) *FilterBuilder {
+	fb.filter.kinds[kind] = EntityFilterFuncAcceptAll
+	return fb
+}
+
+// AddKindWithEntityFilter adds an entity kind with an associated entity filter function.
+// The built filter will match all entities of the added kind for which the entity filter function returns true.
+func (fb *FilterBuilder) AddKindWithEntityFilter(kind Kind, entityFilterFunc EntityFilterFunc) *FilterBuilder {
+	fb.filter.kinds[kind] = entityFilterFunc
+	return fb
 }
 
-// MatchKind returns true if the filter matches the passed Kind. If the filter
-// is nil, or has no kinds, it always matches.
-func (f *Filter) MatchKind(k Kind) bool {
+// SetSource sets the source for the filter.
+func (fb *FilterBuilder) SetSource(source Source) *FilterBuilder {
+	fb.filter.source = source
+	return fb
+}
+
+// SetEventType sets the event type for the filter
+func (fb *FilterBuilder) SetEventType(eventType EventType) *FilterBuilder {
+	fb.filter.eventType = eventType
+	return fb
+}
+
+// MatchEntity returns true if the filter matches the passed entity.
+// If the filter is nil, or has no kinds, it always matches.
+func (f *Filter) MatchEntity(entity *Entity) bool {
 	if f == nil || len(f.kinds) == 0 {
 		return true
 	}
 
-	_, ok := f.kinds[k]
+	if entity == nil {
+		return false
+	}
+
+	entityKind := (*entity).GetID().Kind
 
-	return ok
+	if entityFilterFunc, found := f.kinds[entityKind]; found {
+		return entityFilterFunc(entity)
+	}
+
+	return false
```

Comment:
At this point it means that there was not a function defined to filter by entity. So shouldn't we accept instead of reject?

Reply:
No I don't think so. 

The `Kinds` field of the filter is a map from acceptable Entity Kind to a an entity filter function.

If the entity kind doesn't exist in the map, it means it should be filtered out (should return false).

Is the code confusing about this point?

Reply:
Ah sorry, you are right :+1: